lib/fdotypes.c

```diff
@@ -5471,6 +5471,110 @@ bool fdo_serviceinfo_modules_list_write(fdow_t *fdow) {
+ * @param module_list - Pointer to the ServiceInfo module list containing all
+ * Device ServiceInfos modules.
+ * @param mtu - MTU to be used for fitting the values.
+ *
+ * @return true if there is ServiceInfo to be sent in the next iteration,
+ * false on any other conditions.
+ */
+bool fdo_serviceinfo_external_mod_is_more(fdow_t *fdow,
+	fdo_sdk_service_info_module_list_t *module_list, size_t mtu) {
+
+	if (!fdow || !module_list) {
+		return false;
+	}
+	fdo_sdk_service_info_module_list_t *traverse_list = module_list;
+	bool is_more = false;
+
+	while (traverse_list) {
+		if (traverse_list->module.active &&
+			traverse_list->module.service_info_callback(
+			FDO_SI_IS_MORE_DSI, NULL, fdow, NULL, NULL, &is_more, mtu) != FDO_SI_SUCCESS) {
+			LOG(LOG_DEBUG, "Sv_info: %s's CB Failed for type:%d\n",
+			    traverse_list->module.module_name, FDO_SI_HAS_MORE_DSI);
+			return false;
+		}
+		if (is_more) {
+			return is_more;
+		}
+		traverse_list = traverse_list->next;
+	}
+	return false;
+}
+
+/**
+ * Return a module reference that has some ServiceInfo to be send NOW/immediately,
+ * by making a callbacks to each active module, to determine whether the module
+ * has something to send immediately.
+ *
+ * @param fdow - Pointer to the writer.
+ * @param module_list - Pointer to the ServiceInfo module list containing all
+ * Device ServiceInfos modules.
+ * @param mtu - MTU to be used for fitting the values.
+ *
+ * @return Pointer to/module reference (fdo_sdk_service_info_module *) if there is any module
+ * that has ServiceInfo to send NOW/immediately, else return NULL.
+ */
+fdo_sdk_service_info_module* fdo_serviceinfo_get_external_mod_to_write(fdow_t *fdow,
+	fdo_sdk_service_info_module_list_t *module_list, size_t mtu) {
+
+	if (!fdow || !module_list) {
+		return false;
+	}
+	fdo_sdk_service_info_module_list_t *traverse_list = module_list;
+	bool has_more = false;
+
+	while (traverse_list) {
+		if (traverse_list->module.active &&
+			traverse_list->module.service_info_callback(
+			FDO_SI_HAS_MORE_DSI, NULL, fdow, NULL, &has_more, NULL, mtu) != FDO_SI_SUCCESS) {
+			LOG(LOG_DEBUG, "Sv_info: %s's CB Failed for type:%d\n",
+			    traverse_list->module.module_name, FDO_SI_HAS_MORE_DSI);
+			return false;
+		}
+		if (has_more) {
+			return &(traverse_list->module);
+		}
+		traverse_list = traverse_list->next;
+	}
+	return NULL;
+}
+
+/**
+ * Given an active ServiceInfo module, invoke the callback on the same,
+ * to write ServiceInfo to be sent. The module is responsible for writing the 'ServiceInfo'
+ * structure.
+ *
+ * @param fdow - Pointer to the writer.
+ * @param module - Pointer to the ServiceInfo module list containing all
+ * Device ServiceInfos modules.
+ * @param mtu - MTU to be used for fitting the values.
+ *
+ * @return Return true if module callback was successful, else return false.
+ */
+bool fdo_serviceinfo_external_mod_write(fdow_t *fdow, fdo_sdk_service_info_module *module,
+	size_t mtu) {
+	bool ret = false;
```

Comment:
Do we need to use the ret variable? It seems that returning true or false in each case is sufficient.

Reply:
Updated.